src/btree/node.ts

```diff
@@ -1,158 +1,139 @@
+		const size = dataView.getUint32(0);
 
-			return { node, bytesRead: currentOffset };
-		} catch (error) {
-			// console.error(error);
-			return { node: null, bytesRead: 0 };
+		const leaf = size < 0;
+
+		if (leaf) {
+			this.leafPointers = Array<MemoryPointer>(-size);
+			this.keys = Array<ReferencedValue>(-size);
+		} else {
+			this.internalPointers = Array<bigint>(size + 1);
+			this.keys = Array<ReferencedValue>(size);
+		}
+
+		if (size === 0) {
+			throw new Error("empty node");
 		}
-	}
 
-	async bsearch(key: Uint8Array): Promise<number> {
-		let lo = 0;
-		let hi = this.keys.length - 1;
-
-		while (lo <= hi) {
-			const mid = Math.floor((lo + hi) / 2);
-			const cmp = compareBytes(key, this.keys[mid].value);
-
-			switch (cmp) {
-				case 0:
-					return mid;
-				case -1:
-					hi = mid - 1;
-					break;
-				case 1:
-					lo = mid + 1;
-					break;
+		let m = 4;
+
+		for (let idx = 0; idx <= this.keys.length - 1; idx++) {
+			dataView = new DataView(buffer.slice(m, m + 4));
+			const l = dataView.getUint32(0);
+			if (l === ~0 >>> 0) {
+				dataView = new DataView(buffer.slice(m + 4, m + 12));
+				this.keys[idx].dataPointer.offset = dataView.getBigUint64(0);
+				dataView = new DataView(buffer.slice(m + 12, m + 16));
+				this.keys[idx].dataPointer.length = dataView.getUint32(0);
+
+				const dp = this.keys[idx].dataPointer;
+				const dataSlice = this.data.slice(
+					Number(dp.offset),
+					Number(dp.offset + BigInt(dp.length))
+				);
+				this.keys[idx].value = new Uint8Array(dataSlice);
+
+				m += 4 + 12;
+			} else {
+				this.keys[idx].value = new Uint8Array(buffer.slice(m + 4, m + 4 + l));
+				m += 4 + l;
 			}
 		}
 
-		return ~lo;
-	}
-}
+		for (let idx = 0; idx <= this.leafPointers.length - 1; idx++) {
+			dataView = new DataView(buffer.slice(m, m + 8));
+			this.leafPointers[idx].offset = dataView.getBigUint64(0);
+			dataView = new DataView(buffer.slice(m + 8, m + 12));
+			this.leafPointers[idx].length = dataView.getUint32(0);
 
-export function compareBytes(a: Uint8Array, b: Uint8Array): number {
-	const len = Math.min(a.length, b.length);
+			m += 12;
+		}
+
+		for (let idx = 0; idx <= this.internalPointers.length - 1; idx++) {
+			dataView = new DataView(buffer.slice(m, m + 8));
+			this.internalPointers[idx] = dataView.getBigUint64(0);
 
-	for (let idx = 0; idx < len; idx++) {
-		if (a[idx] !== b[idx]) {
-			return a[idx] < b[idx] ? -1 : 1;
+			m += 8;
 		}
 	}
 
-	if (a.length < b.length) {
-		return -1;
-	}
-	if (a.length > b.length) {
-		return 1;
+	static async fromMemoryPointer(
+		mp: MemoryPointer,
+		resolver: RangeResolver,
+		data: ArrayBuffer
+	): Promise<{ node: BPTreeNode; bytesRead: number }> {
+		const { data: bufferData } = await resolver({
+			start: Number(mp.offset),
+			end: Number(mp.offset) + 4096 - 1,
+		});
+		const node = new BPTreeNode([], [], [], data);
+		await node.unmarshalBinary(bufferData);
+
+		return { node, bytesRead: 4096 };
```

Comment:
We pass in the `data: ArrayBuffer` from the `bptree`. Now that I think about it, since this is array is for our values, this should be a `Uint8Array`.

Also for `bytesRead`, I should probably return `m` from `unmarshalBinary`